hs_core/views/utils.py

```diff
@@ -46,31 +59,58 @@ def upload_from_irods(username, password, host, port, zone, irods_fnames, res_fi
+       b. add file to resource
+       c. delete a file from the resource
+       authorize(request, res_id=id_of_resource, needed_permission=Action_To_Authorize.EDIT_RESOURCE)
+    4. User wants to set resource flag (public, published, shareable etc):
+       authorize(request, res_id=id_of_resource, needed_permission=Action_To_Authorize.SET_RESOURCE_FLAG)
+    5. User wants to delete a resource:
+       authorize(request, res_id=id_of_resource, needed_permission=Action_To_Authorize.DELETE_RESOURCE)
+    6. User wants to create new version of a resource:
+       authorize(request, res_id=id_of_resource, needed_permission=Action_To_Authorize.CREATE_RESOURCE_VERSION)
+
+    Note: resource 'shareable' status has no effect on authorization
     """
+    authorized = False
     user = get_user(request)
-    if res is None:
-        try:
-            res = hydroshare.utils.get_resource_by_shortkey(res_id, or_404=False)
-        except ObjectDoesNotExist:
-            raise NotFound(detail="No resource was found for resource id:%s" % res_id)
-
-    if discoverable and (edit is False and view is False and full is False and superuser is False):
-        authorized = res.raccess.discoverable
-    elif user.is_authenticated():
-        authorized = (view and user.uaccess.can_view_resource(res)) or \
-                     (edit and user.uaccess.can_change_resource(res)) or \
-                     (full and user.uaccess.can_delete_resource(res)) or \
-                     (discoverable and res.raccess.discoverable) or \
-                     (superuser and user.is_superuser)
-    else:
-        authorized = (view and res.raccess.public) or \
-                     (discoverable and res.raccess.discoverable)
-                     
+
+    try:
+        res = hydroshare.utils.get_resource_by_shortkey(res_id, or_404=False)
+    except ObjectDoesNotExist:
+        raise NotFound(detail="No resource was found for resource id:%s" % res_id)
+
+    if needed_permission == Action_To_Authorize.VIEW_METADATA:
+        if res.raccess.discoverable or res.raccess.public:
+            authorized = True
+        elif user.is_authenticated() and user.is_active:
+            authorized = user.uaccess.can_view_resource(res)
+    elif user.is_authenticated() and user.is_active:
+        if needed_permission == Action_To_Authorize.VIEW_RESOURCE:
+            authorized = user.uaccess.can_view_resource(res)
+        elif needed_permission == Action_To_Authorize.EDIT_RESOURCE:
+            authorized = user.uaccess.can_change_resource(res)
+        elif needed_permission == Action_To_Authorize.DELETE_RESOURCE:
+            authorized = user.uaccess.can_delete_resource(res)
+        elif needed_permission == Action_To_Authorize.SET_RESOURCE_FLAG:
+            authorized = user.uaccess.can_change_resource_flags(res)
+        elif needed_permission == Action_To_Authorize.CREATE_RESOURCE_VERSION:
+            authorized = user.uaccess.can_create_resource_version(res)
```

Comment:
@pkdash Can we just use user.uaccess.owns_resource() to authorize CREATE_RESOURCE_VERSION action?